platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/projectmodule/ProjectPublicationRegistry.java

```diff
@@ -16,35 +16,48 @@
 
 package org.gradle.api.internal.artifacts.ivyservice.projectmodule;
 
-import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.artifacts.component.BuildIdentifier;
+import org.gradle.api.internal.project.ProjectIdentity;
 import org.gradle.util.Path;
 
 import javax.annotation.concurrent.ThreadSafe;
 import java.util.Collection;
 
 /**
- * A build scoped service that collects information on the local "publications" of each project within a build. A "publication" here means some buildable thing that the project produces that can be consumed outside of the project.
+ * A build tree scoped service that collects information on the local "publications" of each
+ * project within a build tree. A "publication" here means some buildable thing that the project
+ * produces that can be consumed outside of the project.
  *
  * The information is gathered from multiple sources ({@code publishing.publications} container, etc.).
  */
 @ThreadSafe
 public interface ProjectPublicationRegistry {
-    void registerPublication(ProjectInternal project, ProjectPublication publication);
+
+    /**
+     * Register the given publication as produced by the project with the given ID.
+     */
+    void registerPublication(ProjectIdentity projectIdentity, ProjectPublication publication);
 
     /**
      * Returns the known publications for the given project.
      */
-    <T extends ProjectPublication> Collection<T> getPublications(Class<T> type, Path projectIdentityPath);
+    <T extends ProjectPublication> Collection<T> getPublicationsForProject(Class<T> type, Path projectIdentityPath);
 
     /**
-     * Returns all known publications.
+     * Returns all known publications for the given build.
      */
-    <T extends ProjectPublication> Collection<Reference<T>> getPublications(Class<T> type);
+    <T extends ProjectPublication> Collection<PublicationForProject<T>> getPublicationsForBuild(Class<T> type, BuildIdentifier buildIdentity);
+
+    interface PublicationForProject<T extends ProjectPublication> {
 
-    interface Reference<T> {
-        T get();
+        /**
+         * The publication produced by the project.
+         */
+        T getPublication();
 
-        // Should use ProjectState instead
```

Comment:
While not strictly necessary to fix the bug, we also enhanced this functionality to not require realized `Project` instances, but to only rely on identifiers